### PR TITLE
Enable restart of signal-interrupted syscalls on x86-64 platform

### LINKED ISSUE #1578: Restart signal-interrupted system calls
### Feature Description

What happens if a POSIX signal is delivered to a thread while it is doing some blocking operation in a system call?

The expected behavior  is one of the two below:
1. Return `EINTR`, or
2. Retry the system call.

The first behavior is trivial to implement. This issue proposes to add the second one.

### Problem or Need

Allow me to quote the man page of [signal(7)](https://man7.org/linux/man-pages/man7/signal.7.html).

> If a signal handler is invoked while a system call or library function call is blocked, then either:
>
> * the call is automatically restarted after the signal handler returns; or
> * the call fails with the error EINTR.
>
> Which of these two behaviors occurs depends on the interface and whether or not the signal handler was established using the SA_RESTART flag (see [sigaction(2)](https://man7.org/linux/man-pages/man2/sigaction.2.html)).  The details vary across UNIX systems; below, the details for Linux.
>
>  If a blocked call to one of the following interfaces is interrupted by a signal handler, then the call is automatically restarted after the signal handler returns if the SA_RESTART flag was used; otherwise the call fails with the error EINTR:

System calls that will be restarted automatically if the `SA_RESTART` flag is set for the signal handlers include
* `read`, `write`, `ioctl`
* `open`
* `wait`, `waitpid`
* file locks via `fnctl`
* `futex`

In particular, we have encountered a CI failure (#1577 ) due to the wrong behavior of signal-interrupted futex system call returning `EINTR`, instead of retrying the system call automatically.

### Suggested Implementation

We need to introduce a general retry mechanism for signal-interrupted system calls. In the case of futex, Linux's futex retry logic can be found [here](https://github.com/torvalds/linux/blob/da4373fbcf006deda90e5e6a87c499e0ff747572/kernel/futex/waitwake.c#L706).

### PR DESCRIPTION
**Pull request by jellllly420**
This PR may close #1578 (though the CI failure mentioned there seems irrelevant to the topic according to the comments) and maybe #1587 (?).

## How this PR implements syscall restart

This PR relies on the interrupted syscall to return a newly introduced errno `ERESTARTSYS`. We check that when we are handling possibly pending signals on our way back to userspace from a syscall; if we see an `ERESTARTSYS`, then we reset the syscall number and set back the instruction pointer by 2 byte (the length of x86-64 `syscall` instruction) therefore we can re-excute the syscall once we go back to userspace.

## Known issues
- the futex implementation may still be wrong since this PR doesn't try to solve the problems discussed in #1578
    - plus, I don't quite like the idea of the `pause_timeout` API. However, I didn't find an appropriate way to get rid of it.
- some syscalls like `getrandom` still doesn't support restart because we currently even don't support their blocking.

### CODE REVIEW

File: kernel/src/process/signal/mod.rs [hunk 1]
```diff
@@ -69,6 +70,18 @@ pub fn handle_pending_signal(user_ctx: &mut UserContext, ctx: &Context) -> Resul
             restorer_addr,
             mask,
         } => {
+            if user_ctx.syscall_num() != usize::MAX
+                && user_ctx.syscall_ret() == -(Errno::ERESTARTSYS as i32) as usize
```
Comment (tatetian) [inline kernel/src/process/signal/mod.rs:74]: **Major issue 1.** We cannot check the conditions this way. The branch gated by these conditions are _only_ for system call handling.  But `handle_pending_signal` may be called after either system call or exception handling. If it is not a system call, then these checks do not make any sense.

Comment (jellllly420) [reply kernel/src/process/signal/mod.rs:78]: Just to put it clear, I added some logics in the assembly and signal-related code so that the rax would be -1 if it's not coming from a syscall. After all, I agree that we should not add more assembly logics unless necessary so I have modified as you advised.

**Comment by jellllly420** [comment]
The second commit https://github.com/asterinas/asterinas/pull/1598/commits/08d4c76cc8d998cb5d494dbcfd96549c420aa205 in this PR reverts #1642. It's because I want the futex_wake to succeed when a futex_wake happens right between a timeout/interrupt of futex_wake and the return of futex_wake. So I have to check whether the futex item is still queued at the end of futex_wake then we can just remove the futex item there instead of deferring the removal. Do you think the change is OK or not? @tangruize

**Comment by tangruize** [comment]
> It's because I want the futex_wait to succeed when a futex_wake happens right between a timeout/interrupt of futex_wait and the return of futex_wait.

**Comment by tangruize** [comment]
> It's because I want the futex_wait to succeed when a futex_wake happens right between a timeout/interrupt of futex_wait and the return of futex_wait.

With this fix, since the `futex_item` is deleted before `futex_wait` returns, it avoids lost wakeups if `remove_and_wake_items` does not check whether `item.wake()` actually woke a waiter. If we continue checking the return value of `item.wake()`, it might lead to spurious wakeups if a `futex_item` is already woken by a signal. Spurious wakeups are allowed in futex, so they wouldn't cause bugs. That said, I prefer to keep the check as a safeguard because future code changes might violate the assumption that no already-woken items remain in the queue.

> So I have to check whether the futex item is still queued at the end of futex_wake then we can just remove the futex item there instead of deferring the removal.

I'm sorry, but I didn't fully understand it...

**Comment by jellllly420** [comment]
@tangruize OK then I'll keep the check as a guard.

> I'm sorry, but I didn't fully understand it...

Sorry I mean at the end of futex_wait (not futex_wake) where I remove the futex_item.

**Comment by jellllly420** [comment]
@boterinas rerun

File: kernel/src/process/posix_thread/futex.rs [hunk 2]
```diff
@@ -68,12 +68,20 @@ pub fn futex_wait_bitset(
...
-    futex_bucket.add_item(futex_item);
+    futex_bucket.add_item(futex_item.clone());
 
     // drop lock
     drop(futex_bucket);
 
-    waiter.pause_timeout(timeout)
+    let res = waiter.pause_timeout(timeout);
+    futex_bucket = futex_bucket_ref.lock();
+    if futex_bucket.remove_item(&futex_item).is_err() {
```
[the remark this answers is not in the document]
Comment (jellllly420) [reply kernel/src/process/posix_thread/futex.rs:78]: I tried to `remove_item` here to handle such a case:
1. Process A futex_wait on a futex var, later got interrupted by a signal, and then
2. Process B tried to futex_wake a process waiting on the futex var, removing the futex_item, and then
3. Process A haven't return from the futex_wait, then what should be the correct return value?

Linux's behavior is to let A's futex_wait succeed, avoiding another unnecessary restart, which can be way more expensive than a `lock` and `remove_item` (It may not be appropriate to compare this way since they don't occur in the same frequency). To achieve that, A should check whether its futex_item is still queued before returning to userspace even if it's interrupted by a signal. So here comes the locking and scanning. Linux does the same thing (https://elixir.bootlin.com/linux/v6.12.1/source/kernel/futex/waitwake.c#L672).

I can get your concern on the possible regression with additional `lock` and `remove_item`. However, I think futex cares a lot more about the performance before its sleeping. When it's woken from a sleep, I mean, a sleep of arbitary length, it would not cares that much about a lock and scan (remove) like it used to. However, I don't have a convincing benchmark to defense my code and I'll keep this issue for future discussion if you'd like that way.

File: kernel/src/process/signal/pause.rs [hunk 3]
```diff
@@ -22,16 +22,18 @@ use crate::{
...
 pub trait Pause: WaitTimeout {
     /// Pauses until the condition is met or a signal interrupts.
     ///
     /// # Errors
     ///
-    /// This method will return an error with [`EINTR`] if a signal is received before the
-    /// condition is met.
+    /// Before the condition is met, this method will return an error with
+    ///  - [`ERESTARTSYS`] if a signal is received and `timeout` is `None`;
+    ///  - [`EINTR`] if a signal is received and `timeout` is `Some(_)`.
```
[the remark this answers is not in the document]
Comment (jellllly420) [reply kernel/src/process/signal/pause.rs:33]: Thank you for the thorough comments! I'd like to reply one by one to see whether we can reach an agreement on whether the changes make sense.

> First, the semantics of the error code is that the system call should be restarted. But this Pause trait is a primitive that is decoupled from any particular system call implementation. So the interface of a Pause object should not make decisions at the system call level.

You are right about all being said but may be wrong with the assumption: this is not about coupling or decoupling things.

The original idea is that a `pause` is like a syscall that can sleep, from a user's perspective (although `pause`'s user is kernel developer and syscall's user is kernel user). They both wait for some events until timing out and may get interrupted, i.e., they are both interruptable waiting.

An interruptable waiting ought to have a mechanism of restarting (no matter what it is, `pause` or syscall) for its caller. The change is not to modify `pause` to adjust to syscall's stuff but to make `pause`'s own semantics complete. It just happens to work well with this PR.

> Second, the comments here are simply invalid. This pause_until method does not even have a timeout argument.

Sorry that's my bad. According to what I stated before, the basic idea is that `pause` should notify its users (here, returning `ERESTARTSYS`) when it can be restarted with exactly the same parameters (including the timer parameter). The comments are not correct here.

> Third, even for pause methods that do have a timeout argument, I don't see why the error code should depend on the input timeout? The caller is fully aware of whether timeout is Some(_) or None. If a Err(EINTR) is returned, the caller can convert the error code, if the conversion is necessary to comply with Linux's behaviors.

It's `pause`'s users' choice whether to restart a `pause`.

You can see it as a syscall returning `EINTR` to its user. At this time the user can choose whether or not to restart. But you cannot just remove the `EINTR` error code from kernel just because some users decide not to restart, which is not fair.

Some syscalls needs special handling logic because they are basically waiting for a signal or doing similar things. They surely don't need a restart since they are just waiting to be interrupted. This is not about a wrong decision being made then got corrected. 

> > I don't quite like the idea of the pause_timeout API. However, I didn't find an appropriate way to get rid of it.

> I don't know your thought process. If you don't like the current Pause interface, fine. Explain why and justify you modifications to the interface. The proposed changes do not make much sense to me at all.

After all I have stated, I have to clarify that I'm not quite against the idea of whole `Pause` trait but the `pause_timeout` method in it. It's just so different from other methods in `Pause` though seems similar. I tried to implement it using `pause_until_or_timeout_impl` (just like any other methods) but didn't find a way.

File: kernel/src/process/posix_thread/futex.rs [hunk 4]
```diff
@@ -68,12 +68,20 @@ pub fn futex_wait_bitset(
         );
     }
 
-    futex_bucket.add_item(futex_item);
+    futex_bucket.add_item(futex_item.clone());
```
[the remark this answers is not in the document]
Comment (jellllly420) [reply kernel/src/process/posix_thread/futex.rs:71]: It'll be corrected in next iteration.

File: kernel/src/process/posix_thread/futex.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/src/process/posix_thread/futex.rs:78]: > I tried to `remove_item` here to handle such a case:
> 
> 1. Process A futex_wait on a futex var, later got interrupted by a signal, and then
> 2. Process B tried to futex_wake a process waiting on the futex var, removing the futex_item, and then
> 3. Process A haven't return from the futex_wait, then what should be the correct return value?
>
> Linux's behavior is to let A's futex_wait succeed, avoiding another unnecessary restart,

Thank you for explaining the motivation behind the proposed change and providing a clear breakdown of the scenario. I appreciate the good intention. However, as I see it, the current proposal aims to optimize the performance of a rare edge case but does so at the expense of degrading performance in a much more common scenario. This trade-off doesn’t seem to work in our favor, does it?

That said, your insight about removing the enqueued futex item at the exit of `futex_wait_bitset` highlights an important issue—but for a different reason: mitigating a potential denial-of-service (DoS) vulnerability caused by memory leakage. Specifically, the enqueued futex item, represented as a `Box<FutexItem>`, may remain undequeued if a futex wait operation is interrupted by a signal or times out. In such cases, the `Box<FutexItem>` would persist in memory, leaving our implementation vulnerable to exploitation by user programs that could repeatedly issue futex wait operations to exhaust kernel memory.

This is a critical vulnerability, though perhaps not an urgent one. Addressing it requires a carefully designed solution that balances correctness, efficiency, and robustness. It’s essential that we approach this fix holistically to avoid introducing new inefficiencies or errors while resolving the underlying issue.

This solution should be addressed as a separate issue and deserves its own dedicated PR. For now, let's focus on enabling the system call restart mechanism in this PR and reserve any non-trivial or potentially controversial changes to futex for future work. And I am looking forward to such a PR and I will be very glad to discuss with you in it.

File: kernel/src/process/signal/pause.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tatetian) [reply kernel/src/process/signal/pause.rs:33]: > The original idea is that a `pause` is like a syscall that can sleep, from a user's perspective (although `pause`'s user is kernel developer and syscall's user is kernel user). They both wait for some events until timing out and may get interrupted, i.e., they are both interruptable waiting.

Ok. I got this analog. But I am afraid that it is misleading. For a system call implementation such as futex, it knows whether or not that the system call itself should be restarted. It can make the right decision about returning `ERESTARTSYS` or another return value.

But `Pause` is a general-purpose primitive. It has many users, including many different system calls, each of which has its own Linux behaviors, including system call restart, to conform with. So there is no way for `Pause` to make the right call about restarting an operation or not.

> An interruptable waiting ought to have a mechanism of restarting (no matter what it is, `pause` or syscall) for its caller. The change is not to modify `pause` to adjust to syscall's stuff but to make `pause`'s own semantics complete.

The current set of error codes of `Pause`, which are `ETIME` and `EINTR`,  are sufficient for the caller to make the decision about whether or not it should restart the operation.

> I'm not quite against the idea of whole `Pause` trait but the `pause_timeout` method in it. It's just so different from other methods in `Pause` though seems similar.

I am open to any ideas about improving the `Pause` trait, including its methods with `timeout` argument. But you need to make concrete arguments about its shortcomings. Describing them as being "so different from" is a vague and weak argument. And this is probably also a topic that is out of the scope of this PR.

[the remark this answers is not in the document]
Comment (jellllly420) [reply kernel/src/process/signal/pause.rs:33]: Sorry indeed for the delay. I've been busy these few days.

TL;DR: I'm OK to go in your way if you insist to enable this feature first. But I still prefer the original version a little more though I can now see that the proposed solutions both have their own pros and cons.

> For a system call implementation like futex, it has the context to decide whether the system call should be restarted.

Surely syscalls are capable of deciding their return values on their own. But why should they if we can have a centralized method with exactly the same semantics? In the current implementation, the `ETIME` and `EINTR` returned by a syscall are also directly derived from the return value of the wrapped `Pause`.

We have to distinguish between two concepts: (1) `Pause` make decisions for syscalls who invoke them (2) `Pause`'s return values are just information they want their callers to know and syscalls (and maybe other `Pause` callers) can decide based on the information. I think we are actually implementing the second.  It looks so much like the first just because (1) `Pause` and syscalls share the same set of error codes and (2) syscalls usually just derive the return value, e.g., `ETIME`.

I guess you may be concerned with the fact that sometimes the syscalls themselves just don't want to return `ERESTARTSYS` even though the `Pause` return so. But it's syscalls' obligation to decide what they want (derive the return value of `Pause` or return some other error codes based on it). Suppose we go in your way: keep `Pause` unmodified and let syscalls decide according to`EINTR` returned by `Pause`, then we will also be faced with the same decision: derive the `EINTR` or return `ERESTARTSYS` if it's needed. It doesn't make things any better.

I checked how Linux handles this just now. It turns out Linux's [`wait_event` family](https://elixir.bootlin.com/linux/v6.12.4/source/include/linux/wait.h#L480), who are also general-purpose and are used just like our `Pause`, choose to return `ERESTARTSYS` when interrupted by a signal. They don't even return `EINTR`, which means they don't distinguish between different kinds of interruptions. But wouldn't it be better if we do?

**Comment by jellllly420** [comment]
@boterinas rerun

File: kernel/src/syscall/open.rs [hunk 5]
```diff
@@ -50,11 +54,17 @@ pub fn sys_openat(
 }
 
 pub fn sys_open(path_addr: Vaddr, flags: u32, mode: u16, ctx: &Context) -> Result<SyscallReturn> {
-    self::sys_openat(AT_FDCWD, path_addr, flags, mode, ctx)
+    self::sys_openat(AT_FDCWD, path_addr, flags, mode, ctx).map_err(|err| match err.error() {
```
Comment (tatetian) [inline kernel/src/syscall/open.rs:57]: In `sys_openat`, `Errno::EINTR` is already converted to `Errno::ERESTARTSYS`.

File: kernel/src/syscall/accept.rs [hunk 6]
```diff
@@ -18,7 +18,14 @@ pub fn sys_accept(
 ) -> Result<SyscallReturn> {
     debug!("sockfd = {sockfd}, sockaddr_ptr = 0x{sockaddr_ptr:x}, addrlen_ptr = 0x{addrlen_ptr:x}");
 
-    let fd = do_accept(sockfd, sockaddr_ptr, addrlen_ptr, Flags::empty(), ctx)?;
+    let fd =
+        do_accept(sockfd, sockaddr_ptr, addrlen_ptr, Flags::empty(), ctx).map_err(
+            |err| match err.error() {
+                // FIXME: `accept` should not be restarted if a timeout has been set on the socket using `setsockopt`.
+                Errno::EINTR => Error::new(Errno::ERESTARTSYS),
```
Comment (tatetian) [inline kernel/src/syscall/accept.rs:25]: Move the errno number conversion into `do_accept` so that it won't be duplicated in both `sys_accept` and `sys_accept4`.

File: kernel/src/syscall/open.rs [hunk 7]
```diff
@@ -50,11 +54,17 @@ pub fn sys_openat(
...
+        Errno::EINTR => Error::new(Errno::ERESTARTSYS),
+        _ => err,
+    })
 }
 
 pub fn sys_creat(path_addr: Vaddr, mode: u16, ctx: &Context) -> Result<SyscallReturn> {
     let flags =
         AccessMode::O_WRONLY as u32 | CreationFlags::O_CREAT.bits() | CreationFlags::O_TRUNC.bits();
-    self::sys_openat(AT_FDCWD, path_addr, flags, mode, ctx)
+    self::sys_openat(AT_FDCWD, path_addr, flags, mode, ctx).map_err(|err| match err.error() {
```
Comment (tatetian) [inline kernel/src/syscall/open.rs:66]: In `sys_openat`, `Errno::EINTR` is already converted to `Errno::ERESTARTSYS`.

**Review by tatetian — COMMENTED** [review]
Good job! The PR looks almost perfect. Just a few minor issues.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
